In [29]:
import requests
import pandas as pd
import json
from pandas.io.json import json_normalize
from ast import literal_eval
import urllib.parse as urlparse
import warnings
warnings.filterwarnings('ignore')


def get_request_query(url, operation, numofrows, pageno, datatype, params, codename, code, serviceKey):
    params = literal_eval(params)
    params = urlparse.urlencode(params)
    request_query = []
    for i in range(len(code)):
        full_url = url + "/" + operation + "?" + "numOfRows=" + numofrows + "&pageNo=" + pageno + "&dataType=" + datatype + "&"+ params + "&" + codename + "=" + code[i] + "&serviceKey=" + serviceKey
    request_query.append(full_url)
        
    return request_query

def loop_request_query(request_query):
    for query in request_query:
        response = requests.get(url=query)
    return response

def json_response(response):
    json_ = response.text
    json_object = json.loads(json_)
    return json_object

def get_dataframe(json_object):
    df = json_normalize(json_object['response']['body']['items']['item'])
    return df


def get_data():
    URL = str(input("URL 주소를 입력하시오: "))
    API_KEY = str(input("인증키를 입력하시오: "))
    OPERATION = str(input("상세기능명을 입력하시오(영문): "))
    NUMOFROWS = str(input("한 페이지에 노출할 결과 수를 입력하시오: "))
    PAGENO = str(input("페이지 번호를 입력하시오: "))
    DATATYPE = str(input("데이터 타입을 입력하시오: "))
    PARAMS_KEY = list(input("항목명을 공백으로 구분하여 입력하시오: ").split())
    PARAMS_VALUE = list(input("항목값을 공백으로 구분하여 입력하시오: ").split())
    PARAMS = str(dict(zip(PARAMS_KEY, PARAMS_VALUE)))
    CODENAME = str(input('지점번호, 지역코드와 관련된 항목명을 입력하시오: '))
    CODE = list(input('지점번호, 지역코드를 입력하시오: ').split())
    
    request_query = get_request_query(URL, OPERATION, NUMOFROWS, PAGENO, DATATYPE, PARAMS, CODENAME, CODE, API_KEY)
    print(request_query)
    return request_query

def main():    
    request_query = get_data()
    
    response = loop_request_query(request_query)
    
    json_object = json_response(response)
    
    df = get_dataframe(json_object)
    
    df_name = input("파일명을 입력하시오: ")
    df.to_csv('data/' + df_name + '.csv', index=False, encoding='cp949')
    df_describe = df.describe()
    df_describe.to_csv('data/' + df_name +'_describe.csv', index=True, encoding='cp949')

In [30]:
main()

URL 주소를 입력하시오: http://apis.data.go.kr/1360000/AsosHourlyInfoService
인증키를 입력하시오: 0VPHiLnleAB93gjILfRfKRQcBuFBbb9XArHRML7HRiBSrkUV4GZxTwOsm%2Fm%2FloyDIFYPsLSoNQFg2j%2FAfxNfnQ%3D%3D
상세기능명을 입력하시오(영문): getWthrDataList
한 페이지에 노출할 결과 수를 입력하시오: 999
페이지 번호를 입력하시오: 1
데이터 타입을 입력하시오: JSON
항목명을 공백으로 구분하여 입력하시오: dataCd dateCd startDt startHh endDt endHh
항목값을 공백으로 구분하여 입력하시오: ASOS HR 20200101 01 20200726 01
지점번호, 지역코드와 관련된 항목명을 입력하시오: stnIds
지점번호, 지역코드를 입력하시오: 108
['http://apis.data.go.kr/1360000/AsosHourlyInfoService/getWthrDataList?numOfRows=999&pageNo=1&dataType=JSON&dataCd=ASOS&dateCd=HR&startDt=20200101&startHh=01&endDt=20200726&endHh=01&stnIds=108&serviceKey=0VPHiLnleAB93gjILfRfKRQcBuFBbb9XArHRML7HRiBSrkUV4GZxTwOsm%2Fm%2FloyDIFYPsLSoNQFg2j%2FAfxNfnQ%3D%3D']
파일명을 입력하시오: test4


http://apis.data.go.kr/1360000/AsosHourlyInfoService

0VPHiLnleAB93gjILfRfKRQcBuFBbb9XArHRML7HRiBSrkUV4GZxTwOsm%2Fm%2FloyDIFYPsLSoNQFg2j%2FAfxNfnQ%3D%3D

getWthrDataList

dataCd dateCd startDt startHh endDt endHh

ASOS HR 20200101 01 20200726 01
